File: app/src/main/java/com/example/josycom/flowoverstack/network/ApiService.java
package com.example.josycom.flowoverstack.network;

import com.example.josycom.flowoverstack.model.AnswerResponse;
import com.example.josycom.flowoverstack.model.QuestionsResponse;
import com.example.josycom.flowoverstack.util.StringConstants;

import java.util.List;

import io.reactivex.Observable;
import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface ApiService {

    @GET(StringConstants.QUESTIONS_END_POINT)
    Call<QuestionsResponse> getQuestionsForAll(
            @Query("page") int page,
            @Query("pagesize") int pageSize,
            @Query("order") String order,
            @Query("sort") String sortCondition,
            @Query("site") String site,
            @Query(value = "filter", encoded = true) String filter);

    @GET(StringConstants.ANSWERS_END_POINT)
    Observable<AnswerResponse> getAnswersToQuestion(
            @Path("question_id") int id,
            @Query("order") String order,
            @Query("sort") String sortCondition,
            @Query("site") String site,
            @Query(value = "filter", encoded = true) String filter);

    @GET("/2.2/search?order=desc&sort=activity&site=stackoverflow&filter=!9Z(-wwYGT")
    Observable<List<QuestionsResponse>> getQuestionsWithTextInTitle(@Query("intitle") String inTitle);
}

File: app/src/main/java/com/josycom/mayorjay/flowoverstack/data/remote/service/ApiService.kt
package com.josycom.mayorjay.flowoverstack.data.remote.service

import com.josycom.mayorjay.flowoverstack.data.remote.model.AnswerResponse
import com.josycom.mayorjay.flowoverstack.data.remote.model.QuestionsResponse
import com.josycom.mayorjay.flowoverstack.data.remote.model.TagsResponse
import com.josycom.mayorjay.flowoverstack.util.AppConstants
import retrofit2.Call
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface ApiService {

    @GET(AppConstants.QUESTIONS_END_POINT)
    suspend fun getQuestionsForAll(
            @Query("page") page: Int,
            @Query("pagesize") pageSize: Int,
            @Query("order") order: String?,
            @Query("sort") sortCondition: String?,
            @Query("site") site: String?,
            @Query("tagged") tagged: String?,
            @Query(value = "filter", encoded = true) filter: String?,
            @Query("key", encoded = true) siteKey: String?): QuestionsResponse

    @GET(AppConstants.ANSWERS_END_POINT)
    suspend fun getAnswersToQuestion(
            @Path("question_id") id: Int,
            @Query("order") order: String?,
            @Query("sort") sortCondition: String?,
            @Query("site") site: String?,
            @Query(value = "filter", encoded = true) filter: String?,
            @Query("key") siteKey: String?): AnswerResponse

    @GET(AppConstants.SEARCH_END_POINT)
    fun getQuestionsWithTextInTitle(
        @Query("intitle") inTitle: String?,
        @Query("page") page: Int,
        @Query("pagesize") pageSize: Int,
        @Query("key") siteKey: String? = AppConstants.API_KEY
    ): Call<QuestionsResponse?>

    @GET(AppConstants.TAGS_END_POINT)
    suspend fun getAllPopularTags(
            @Query("page") page: Int,
            @Query("pagesize") pageSize: Int,
            @Query("inname") inName: String?,
            @Query("key") siteKey: String?): TagsResponse
}